modules/phase_field/src/postprocessors/FeatureFloodCount.C

```diff
@@ -474,74 +511,164 @@ FeatureFloodCount::mergeSets(bool use_periodic_boundary_info)
+//                        << (physical_intersection ? " physical intersection.\n" : " virtual intersection.\n");
+//              _console << *it1 << *it2;
+//              // DEBUGGING
+
+              it1->_ghosted_ids.swap(set_union);
+              it2->_ghosted_ids.clear();
+
               set_union.clear();
               std::set_union(it1->_periodic_nodes.begin(), it1->_periodic_nodes.end(), it2->_periodic_nodes.begin(), it2->_periodic_nodes.end(), set_union_inserter);
-              it2->_periodic_nodes = set_union;
+              it1->_periodic_nodes.swap(set_union);
+              it2->_periodic_nodes.clear();
+
+              set_union.clear();
+              std::set_union(it1->_local_ids.begin(), it1->_local_ids.end(), it2->_local_ids.begin(), it2->_local_ids.end(), set_union_inserter);
+              it1->_local_ids.swap(set_union);
+              it2->_local_ids.clear();
+
+              set_union.clear();
+              std::set_union(it1->_halo_ids.begin(), it1->_halo_ids.end(), it2->_halo_ids.begin(), it2->_halo_ids.end(), set_union_inserter);
+              it1->_halo_ids.swap(set_union);
+              it2->_halo_ids.clear();
+
+              /**
+               * If we had a physical intersection, we need to expand boxes. If we had a virtual (periodic) intersection we need to preserve
+               * all of the boxes from each of the regions' sets.
+               */
+              if (physical_intersection)
+                it1->expandBBox(*it2);
+              else
+                std::copy(it2->_bboxes.begin(), it2->_bboxes.end(), std::back_inserter(it1->_bboxes));
+
+              // Update the min feature id
+              it1->_min_entity_id = std::min(it1->_min_entity_id, it2->_min_entity_id);
+
+              // Set the flag on the merged set so we don't revisit it again
+              it2->_merged = true;
+
+              // Something was merged so we'll need to repeat this loop
+              region_merged = true;
             }
+          } // it2 loop
 
-            // Now remove the merged set, the one we didn't update (it1)
-            _bubble_sets[map_num].erase(it1++);
-            // don't increment the outer loop since we just deleted it incremented
-            need_it1_increment = false;
-            // break out of the inner loop and move on
-            break;
-          }
-        }
+          // Don't increment if we had a merge, we need to retry earlier candidates again
+          if (!region_merged)
+            ++it1;
+          else
+            goto REPEAT_MERGE_LOOPS;
+
+        } // it1 loop
+      } // rank2 loop
+    } // rank1 loop
+  } // map loop
+
+
+  for (unsigned int map_num = 0; map_num < _maps_size; ++map_num)
+    _feature_sets[map_num].clear();
 
-        if (need_it1_increment)
-          ++it1;
+  // Now consolidate the data structure
+  _feature_count = 0;
+  for (processor_id_type rank = 0; rank < n_procs; ++rank)
+  {
```

Comment:
Sometimes I don't like the look of leaving the parens off when there's deep nesting that does have parens. I did omit them in a few places but this particular one I did not.